pkg/cloudinit/cloudinit.go

```diff
@@ -0,0 +1,68 @@
+package cloudinit
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	cloudinitv1 "github.com/harvester/node-manager/pkg/apis/node.harvesterhci.io/v1beta1"
+)
+
+const AnnotationHash = "node.harvesterhci.io/cloudinit-hash"
+
+var Directory = "/host/oem"
+
+// RequireLocal ensures that the Elemental cloud-init file described by
+// the given `cloudinit` object is an exact copy of the `cloudinit` object's
+// contents.
+func RequireLocal(cloudinit *cloudinitv1.CloudInit) error {
+	absPath := filepath.Join(Directory, cloudinit.Spec.Filename)
+
+	f, err := os.Open(absPath)
+	var r io.Reader = f
+	if err != nil {
+		r = strings.NewReader("")
+	} else {
+		defer f.Close()
+	}
+
+	diskChecksum, err := Measure(r)
+	if err != nil {
+		return err
+	}
+
+	if fmt.Sprintf("%x", diskChecksum) == cloudinit.Annotations[AnnotationHash] {
+		return nil
+	}
+
+	tempFile, err := os.CreateTemp(Directory, "node-manager")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempFile.Name())
```

Comment:
should those 2 still be deferred after `os.Rename` ?

 `tempFile.Close()`  and then `os.Rename` ?

Reply:
It is harmless during the success path and it helps tidy things up if the io.Copy fails.